### ISSUE TITLE
getDynamicObjectRotation

### ISSUE THREAD

**Issue by ThomasNightfire** [multitheftauto/mtasa-blue#2340]
**Is your feature request related to a problem? Please describe.**
Tried to get rotation of serverside dynamic object with ID 2968, and got [0, 0, 0].

**Describe the solution you'd like**
A new function: _float float float getDynamicObjectRotation(object obj)_

**Comment by MegadreamsBE** [comment]
Any reason why it has to be a separate function rather than a fix to getElementRotation?

**Comment by PlatinMTA** [comment]
With dynamic objects you are talking about the ones with physics right?

Afaik those arent synced with the server, because the server doesnt understand RenderWave or the gtasa engine on a real level, unlike goldsource or source for example.

Because of that, the only way to get the real rotation of a dynamic object serverside would be to have a syncer that reports the rotation every few ticks or when the function is executed. I dont know how worth it it really is tho.

You should be able to get the rotation of the object clientside, but you would soon notice that the objects movements are pretty wacky and erratic, they tend to get out of sync almost instantly after touching them.

What you need to do to fix this issue on your own is to make a custom syncer for the thing you are trying to do, if any. I made a basketball minigame a while ago and I used a dynamic object, but this object was synced by the one that had thrown the ball (or had it in its hands atm) so even if there was a lot of desync (which there was) it would come back to sync after certain criteria for syncing was complete (someone scored, the ball was dropped or stop moving, etc.)

**Comment by ThomasNightfire** [comment]
> With dynamic objects you are talking about the ones with physics right?
> 
> Afaik those arent synced with the server, because the server doesnt understand RenderWave or the gtasa engine on a real level, unlike goldsource or source for example.
> 
> Because of that, the only way to get the real rotation of a dynamic object serverside would be to have a syncer that reports the rotation every few ticks or when the function is executed. I dont know how worth it it really is tho.
> 
> You should be able to get the rotation of the object clientside, but you would soon notice that the objects movements are pretty wacky and erratic, they tend to get out of sync almost instantly after touching them.
> 
> What you need to do to fix this issue on your own is to make a custom syncer for the thing you are trying to do, if any. I made a basketball minigame a while ago and I used a dynamic object, but this object was synced by the one that had thrown the ball (or had it in its hands atm) so even if there was a lot of desync (which there was) it would come back to sync after certain criteria for syncing was complete (someone scored, the ball was dropped or stop moving, etc.)

Unable to get real rotation of dynamic object on clientside too. It always returns (0, 0, 0) unless rotation is set manually.
About ur solution for sync: I've already done it, but faced this problem with getting rotation; it seems there is no way to get real rotation of 'object with physics' after player/other element touches it, it still return last manually set rotation (including creation rotation of course).

**Comment by PlatinMTA** [comment]
> Unable to get real rotation of dynamic object on clientside too. It always returns (0, 0, 0) unless rotation is set manually.
> About ur solution for sync: I've already done it, but faced this problem with getting rotation; it seems there is no way to get real rotation of 'object with physics' after player/other element touches it, it still return last manually set rotation (including creation rotation of course).

You are right, getElementRotation() gives an incorrect value. The element matrix, however, works just fine.

```lua
local pizza = createObject(1582, 0, 0, 5)

function onRender()
	local x, y, z = getElementPosition(pizza)
	local rx, ry, rz = getElementRotation(pizza)
	
	local mat = getElementMatrix(pizza)
	
	dxDrawText(x..", "..y..", "..z.."\n"..rx..", "..ry..", "..rz, 800, 600)
	
	local xx, xy, xz = getPositionFromElementMatrixOffset(mat, 1, 0, 0)
	local yx, yy, yz = getPositionFromElementMatrixOffset(mat, 0, 1, 0)
	local zx, zy, zz = getPositionFromElementMatrixOffset(mat, 0, 0, 1)
	dxDrawLine3D(x, y, z, xx, xy, xz, tocolor(200, 0, 0, 200), 5.0)
	dxDrawLine3D(x, y, z, yx, yy, yz, tocolor(0, 200, 0, 200), 5.0)
	dxDrawLine3D(x, y, z, zx, zy, zz, tocolor(0, 0, 200, 200), 5.0)
end
addEventHandler("onClientRender", root, onRender)

function getPositionFromElementMatrixOffset(m,offX,offY,offZ) --https://wiki.multitheftauto.com/wiki/GetElementMatrix
    local x = offX * m[1][1] + offY * m[2][1] + offZ * m[3][1] + m[4][1]  -- Apply transform
    local y = offX * m[1][2] + offY * m[2][2] + offZ * m[3][2] + m[4][2]
    local z = offX * m[1][3] + offY * m[2][3] + offZ * m[3][3] + m[4][3]
    return x, y, z                               -- Return the transformed point
end
```

https://youtu.be/bVkWRSo_K60 (first row of numbers X,Y,Z position, second row of numbers RX,RY,RZ)

In the mean time, you could get the rotation from the object from its Matrix, which is not the best of solutions, but should be very straight forward.

**Comment by Einheit-101** [comment]
This is obviously a bug because getElementRotation should take this into account.

**Comment by patrikjuvonen** [comment]
> This is obviously a bug because getElementRotation should take this into account.

Should it though? What about synchronization from server to client? The values would be different and is confusing.

**Comment by ThomasNightfire** [comment]
> > Unable to get real rotation of dynamic object on clientside too. It always returns (0, 0, 0) unless rotation is set manually.
> > About ur solution for sync: I've already done it, but faced this problem with getting rotation; it seems there is no way to get real rotation of 'object with physics' after player/other element touches it, it still return last manually set rotation (including creation rotation of course).
> 
> You are right, getElementRotation() gives an incorrect value. The element matrix, however, works just fine.
> 
> ```lua
> local pizza = createObject(1582, 0, 0, 5)
> 
> function onRender()
> 	local x, y, z = getElementPosition(pizza)
> 	local rx, ry, rz = getElementRotation(pizza)
> 	
> 	local mat = getElementMatrix(pizza)
> 	
> 	dxDrawText(x..", "..y..", "..z.."\n"..rx..", "..ry..", "..rz, 800, 600)
> 	
> 	local xx, xy, xz = getPositionFromElementMatrixOffset(mat, 1, 0, 0)
> 	local yx, yy, yz = getPositionFromElementMatrixOffset(mat, 0, 1, 0)
> 	local zx, zy, zz = getPositionFromElementMatrixOffset(mat, 0, 0, 1)
> 	dxDrawLine3D(x, y, z, xx, xy, xz, tocolor(200, 0, 0, 200), 5.0)
> 	dxDrawLine3D(x, y, z, yx, yy, yz, tocolor(0, 200, 0, 200), 5.0)
> 	dxDrawLine3D(x, y, z, zx, zy, zz, tocolor(0, 0, 200, 200), 5.0)
> end
> addEventHandler("onClientRender", root, onRender)
> 
> function getPositionFromElementMatrixOffset(m,offX,offY,offZ) --https://wiki.multitheftauto.com/wiki/GetElementMatrix
>     local x = offX * m[1][1] + offY * m[2][1] + offZ * m[3][1] + m[4][1]  -- Apply transform
>     local y = offX * m[1][2] + offY * m[2][2] + offZ * m[3][2] + m[4][2]
>     local z = offX * m[1][3] + offY * m[2][3] + offZ * m[3][3] + m[4][3]
>     return x, y, z                               -- Return the transformed point
> end
> ```
> 
> https://youtu.be/bVkWRSo_K60 (first row of numbers X,Y,Z position, second row of numbers RX,RY,RZ)
> 
> In the mean time, you could get the rotation from the object from its Matrix, which is not the best of solutions, but should be very straight forward.

Thanks, man. Didn't event thought about element matrix method, I will use it.
And what's wrong with this solution by the way?

**Comment by PlatinMTA** [comment]
> Thanks, man. Didn't event thought about element matrix method, I will use it.
> And what's wrong with this solution by the way?

By using the element matrix? It's just more expensive to calculate afaik. Having a clientside function made in C that can give you the RotX, RotY and RotZ of the object should be faster.
